app/src/test/kotlin/com/hedvig/app/feature/genericauth/OtpInputViewModelTest.kt

```diff
@@ -1,192 +1,159 @@
-        TODO("Not yet implemented")
-      }
-
-    },
-    uploadMarketAndLanguagePreferencesUseCase = mockk(relaxed = true),
-  )
+  @get:Rule
+  val testFolder = TemporaryFolder()
 
   @Test
   fun testNetworkError() = runTest {
-    otpResult = SubmitOtpResult.Error("Error")
+    val authRepository = FakeAuthRepository()
+    val viewModel = testViewModel(authRepository)
 
     viewModel.submitCode("123456")
-
+    runCurrent()
     assertThat(viewModel.viewState.value.loadingCode).isEqualTo(true)
     assertThat(viewModel.viewState.value.networkErrorMessage).isEqualTo(null)
 
-    advanceUntilIdle()
+    authRepository.submitOtpResponse.add(SubmitOtpResult.Error("Error"))
+    runCurrent()
     assertThat(viewModel.viewState.value.loadingCode).isEqualTo(false)
     assertThat(viewModel.viewState.value.networkErrorMessage).isEqualTo("Error")
   }
 
   @Test
   fun testDismissNetworkError() = runTest {
-    otpResult = SubmitOtpResult.Error("Error")
+    val authRepository = FakeAuthRepository()
+    val viewModel = testViewModel(authRepository)
+    val errorMessage = "Error"
 
     viewModel.submitCode("123456")
+    runCurrent()
     assertThat(viewModel.viewState.value.loadingCode).isEqualTo(true)
     assertThat(viewModel.viewState.value.loadingResend).isEqualTo(false)
 
-    advanceUntilIdle()
-    assertThat(viewModel.viewState.value.networkErrorMessage).isEqualTo("Error")
+    authRepository.submitOtpResponse.add(SubmitOtpResult.Error(errorMessage))
+    runCurrent()
+    assertThat(viewModel.viewState.value.networkErrorMessage).isEqualTo(errorMessage)
     viewModel.dismissError()
     assertThat(viewModel.viewState.value.networkErrorMessage).isEqualTo(null)
   }
 
   @Test
   fun testOtpError() = runTest {
-    otpResult = SubmitOtpResult.Error("Error")
+    val authRepository = FakeAuthRepository()
+    val viewModel = testViewModel(authRepository)
 
     viewModel.submitCode("123456")
+    runCurrent()
     assertThat(viewModel.viewState.value.loadingCode).isEqualTo(true)
     assertThat(viewModel.viewState.value.loadingResend).isEqualTo(false)
 
-    advanceUntilIdle()
+    authRepository.submitOtpResponse.add(SubmitOtpResult.Error(""))
+    runCurrent()
     assertThat(viewModel.viewState.value.networkErrorMessage).isNotNull()
   }
 
   @Test
   fun testOtpSuccess() = runTest {
-    otpResult = SubmitOtpResult.Success(LoginAuthorizationCode("testOtpSuccess"))
+    val authRepository = FakeAuthRepository()
+    val authTokenStorage = testAuthTokenStorage()
+    val viewModel = testViewModel(authRepository, authTokenStorage)
 
-    val events = mutableListOf<OtpInputViewModel.Event>()
-    val eventCollectingJob = launch {
-      viewModel.events.collect { events.add(it) }
+    authTokenStorage.getTokens().first().also { (resultAccessToken, resultRefreshToken) ->
+      assertThat(resultAccessToken?.token).isNull()
+      assertThat(resultRefreshToken?.token).isNull()
+    }
+    viewModel.events.test {
```

Comment:
So the .test {} function coming from [Turbine](https://github.com/cashapp/turbine) is great because it takes care of what we had to do manually here before.
It ensures that by the end of it, no more events were made other than the ones we observed using `awaitItem()`, so we don't need to check for size == 1 in a list we're populating ourselves.
And it makes sure that at the end it cancels the collection, we don't need to do .cancel on the event collecting job.

Reply:
Cool!